File: FBControlCore/Applications/FBInstalledApplication.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBInstalledApplication.h"

#import "FBBundleDescriptor.h"

static NSString *const FBApplicationInstallTypeStringUnknown = @"unknown";
static NSString *const FBApplicationInstallTypeStringSystem = @"system";
static NSString *const FBApplicationInstallTypeStringMac = @"mac";
static NSString *const FBApplicationInstallTypeStringUser = @"user";
static NSString *const FBApplicationInstallTypeStringUserEnterprise = @"user_enterprise";
static NSString *const FBApplicationInstallTypeStringUserDevelopment = @"user_development";

FBApplicationInstallInfoKey const FBApplicationInstallInfoKeyApplicationType = @"ApplicationType";
FBApplicationInstallInfoKey const FBApplicationInstallInfoKeyBundleIdentifier = @"CFBundleIdentifier";
FBApplicationInstallInfoKey const FBApplicationInstallInfoKeyBundleName = @"CFBundleName";
FBApplicationInstallInfoKey const FBApplicationInstallInfoKeyPath = @"Path";
FBApplicationInstallInfoKey const FBApplicationInstallInfoKeySignerIdentity = @"SignerIdentity";

@implementation FBInstalledApplication

#pragma mark Initializers

+ (instancetype)installedApplicationWithBundle:(FBBundleDescriptor *)bundle installType:(FBApplicationInstallType)installType dataContainer:(NSString *)dataContainer
{
  return [[self alloc] initWithBundle:bundle installType:installType dataContainer:dataContainer];
}

+ (instancetype)installedApplicationWithBundle:(FBBundleDescriptor *)bundle installTypeString:(NSString *)installTypeString signerIdentity:(NSString *)signerIdentity dataContainer:(NSString *)dataContainer
{
  return [self installedApplicationWithBundle:bundle installType:[self installTypeFromString:installTypeString signerIdentity:signerIdentity] dataContainer:dataContainer];
}

- (instancetype)initWithBundle:(FBBundleDescriptor *)bundle installType:(FBApplicationInstallType)installType dataContainer:(NSString *)dataContainer
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _bundle = bundle;
  _installType = installType;
  _dataContainer = dataContainer;

  return self;
}

#pragma mark NSObject

- (NSUInteger)hash
{
  return self.bundle.hash ^ self.installType;
}

- (BOOL)isEqual:(FBInstalledApplication *)object
{
  if (![object isKindOfClass:FBInstalledApplication.class]) {
    return NO;
  }
  return [self.bundle isEqual:object.bundle]
      && self.installType == object.installType
      && (self.dataContainer == object.dataContainer || [self.dataContainer isEqualToString:object.dataContainer]);
}

- (NSString *)description
{
  return [NSString stringWithFormat:
    @"Bundle %@ | Install Type %@ | Container %@",
    self.bundle.description,
    self.installTypeString,
    self.dataContainer
  ];
}

#pragma mark NSCopying

- (instancetype)copyWithZone:(NSZone *)zone
{
  return self;
}

#pragma mark Properties

- (NSString *)installTypeString
{
  return [FBInstalledApplication stringFromApplicationInstallType:self.installType];
}

#pragma mark Install Type

+ (NSString *)stringFromApplicationInstallType:(FBApplicationInstallType)installType
{
  switch (installType) {
    case FBApplicationInstallTypeUser:
      return FBApplicationInstallTypeStringUser;
    case FBApplicationInstallTypeUserDevelopment:
      return FBApplicationInstallTypeStringUserDevelopment;
    case FBApplicationInstallTypeUserEnterprise:
      return FBApplicationInstallTypeStringUserEnterprise;
    case FBApplicationInstallTypeSystem:
      return FBApplicationInstallTypeStringSystem;
    case FBApplicationInstallTypeMac:
      return FBApplicationInstallTypeStringMac;
    default:
      return FBApplicationInstallTypeStringUnknown;
  }
}

+ (FBApplicationInstallType)installTypeFromString:(nullable NSString *)installTypeString signerIdentity:(nullable NSString *)signerIdentity
{
  if (!installTypeString) {
    return FBApplicationInstallTypeUnknown;
  }
  installTypeString = [installTypeString lowercaseString];
  if ([installTypeString isEqualToString:FBApplicationInstallTypeStringSystem]) {
    return FBApplicationInstallTypeSystem;
  }
  if ([installTypeString isEqualToString:FBApplicationInstallTypeStringUser]) {
    if ([signerIdentity containsString:@"iPhone Distribution"]) {
      return FBApplicationInstallTypeUserEnterprise;
    } else if ([signerIdentity containsString:@"iPhone Developer"] || [signerIdentity containsString:@"Apple Development"]) {
      return FBApplicationInstallTypeUserDevelopment;
    }
    return FBApplicationInstallTypeUser;
  }
  if ([installTypeString isEqualToString:FBApplicationInstallTypeStringMac]) {
    return FBApplicationInstallTypeMac;
  }
  return FBApplicationInstallTypeUnknown;
}

@end

File: FBControlCore/Applications/FBInstalledApplication.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

private let installTypeStringUnknown = "unknown"
private let installTypeStringSystem = "system"
private let installTypeStringMac = "mac"
private let installTypeStringUser = "user"
private let installTypeStringUserEnterprise = "user_enterprise"
private let installTypeStringUserDevelopment = "user_development"

/// How an application came to be installed.
@objc public enum ApplicationInstallType: UInt, Sendable {
  /// The Application is unknown.
  case unknown = 0
  /// The Application is part of the Operating System.
  case system = 1
  /// The Application is part of macOS.
  case mac = 2
  /// The Application has been installed by the user.
  case user = 3
  /// The Application has been installed by the user and signed with a distribution certificate.
  case userEnterprise = 4
  /// The Application has been installed by the user and signed with a development certificate.
  case userDevelopment = 5
}

/// Keys of the application info dictionary.
public struct ApplicationInstallInfoKey: RawRepresentable, Hashable, Sendable {
  public let rawValue: String
  public init(rawValue: String) {
    self.rawValue = rawValue
  }

  public static let applicationType = ApplicationInstallInfoKey(rawValue: "ApplicationType")
  public static let bundleIdentifier = ApplicationInstallInfoKey(rawValue: "CFBundleIdentifier")
  public static let bundleName = ApplicationInstallInfoKey(rawValue: "CFBundleName")
  public static let path = ApplicationInstallInfoKey(rawValue: "Path")
  public static let signerIdentity = ApplicationInstallInfoKey(rawValue: "SignerIdentity")
}

public struct FBInstalledApplication: Hashable, Sendable, CustomStringConvertible {

  public let bundle: FBBundleDescriptor
  public let installType: ApplicationInstallType
  public let dataContainer: String?

  public var installTypeString: String {
    FBInstalledApplication.string(from: installType)
  }

  public static func installedApplication(withBundle bundle: FBBundleDescriptor, installType: ApplicationInstallType, dataContainer: String?) -> FBInstalledApplication {
    FBInstalledApplication(bundle: bundle, installType: installType, dataContainer: dataContainer)
  }

  public static func installedApplication(withBundle bundle: FBBundleDescriptor, installTypeString: String?, signerIdentity: String?, dataContainer: String?) -> FBInstalledApplication {
    FBInstalledApplication(bundle: bundle, installTypeString: installTypeString, signerIdentity: signerIdentity, dataContainer: dataContainer)
  }

  public init(bundle: FBBundleDescriptor, installType: ApplicationInstallType, dataContainer: String?) {
    self.bundle = bundle
    self.installType = installType
    self.dataContainer = dataContainer
  }

  public init(bundle: FBBundleDescriptor, installTypeString: String?, signerIdentity: String?, dataContainer: String?) {
    let installType = FBInstalledApplication.installType(from: installTypeString, signerIdentity: signerIdentity)
    self.init(bundle: bundle, installType: installType, dataContainer: dataContainer)
  }

  /// The data container takes part in equality but not in the hash, so two applications
  /// that differ only by their container are unequal and share a hash bucket.
  public func hash(into hasher: inout Hasher) {
    hasher.combine(bundle)
    hasher.combine(installType)
  }

  public static func == (lhs: FBInstalledApplication, rhs: FBInstalledApplication) -> Bool {
    lhs.bundle == rhs.bundle
      && lhs.installType == rhs.installType
      && lhs.dataContainer == rhs.dataContainer
  }

  public var description: String {
    "Bundle \(bundle.description) | Install Type \(installTypeString) | Container \(dataContainer ?? "nil")"
  }

  // MARK: - Install Type Mapping

  public static func string(from installType: ApplicationInstallType) -> String {
    switch installType {
    case .user: return installTypeStringUser
    case .userDevelopment: return installTypeStringUserDevelopment
    case .userEnterprise: return installTypeStringUserEnterprise
    case .system: return installTypeStringSystem
    case .mac: return installTypeStringMac
    case .unknown: return installTypeStringUnknown
    @unknown default: return installTypeStringUnknown
    }
  }

  public static func installType(from installTypeString: String?, signerIdentity: String?) -> ApplicationInstallType {
    guard let installTypeString = installTypeString?.lowercased() else {
      return .unknown
    }
    if installTypeString == installTypeStringSystem {
      return .system
    }
    if installTypeString == installTypeStringUser {
      if let signerIdentity {
        if signerIdentity.contains("iPhone Distribution") {
          return .userEnterprise
        } else if signerIdentity.contains("iPhone Developer") || signerIdentity.contains("Apple Development") {
          return .userDevelopment
        }
      }
      return .user
    }
    if installTypeString == installTypeStringMac {
      return .mac
    }
    return .unknown
  }
}
